src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1374,23 +1375,38 @@ private static boolean isZip64ExtBlockSizeValid(int blockSize) {
         private int[] table;                 // Hash chain heads: indexes into entries
         private int tablelen;                // number of hash heads
 
+        /**
+         * A class representing a key to a ZIP file. A key is based
+         * on the file key if available, or the path value if the
+         * file key is not available. The key is also based on the
+         * file's last modified time to allow for cases where a ZIP
+         * file is re-opened after it has been modified.
+         */
         private static class Key {
-            BasicFileAttributes attrs;
+            final BasicFileAttributes attrs;
             File file;
+            final boolean utf8;
 
-            public Key(File file, BasicFileAttributes attrs) {
+            public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
                 this.attrs = attrs;
                 this.file = file;
+                this.utf8 = zc.isUTF8();
             }
 
             public int hashCode() {
-                long t = attrs.lastModifiedTime().toMillis();
-                return ((int)(t ^ (t >>> 32))) + file.hashCode();
+                long t = utf8 ? 0 : Long.MAX_VALUE;
+                t += attrs.lastModifiedTime().toMillis();
+                Object fk = attrs.fileKey();
+                return Long.hashCode(t) +
+                        (fk != null ? fk.hashCode() : file.hashCode());
             }
 
             public boolean equals(Object obj) {
                 if (obj instanceof Key) {
-                    Key key = (Key)obj;
+                    Key key = (Key) obj;
+                    if (key.utf8 != utf8) {
+                        return false;
+                    }
                     if (!attrs.lastModifiedTime().equals(key.attrs.lastModifiedTime())) {
```

Comment:
The `private static class Key` class does not have the `ZipCoder zc` constructor parameter, here we add it based on current latest code in https://github.com/openjdk/jdk/

Reply:
The backported change does not add the zc argumenht to the call in lin 1436.
Can you please explain why you add it?  Originally it was added for larger change JDK-8243254

Reply:
@GoeLin , 
- Because in the Line `1436` of the same file, it needs the 3rd parameter `zc` . 
  - Without this parameter, there will be a compile error. 
- Considering this is an internal `private` class (line `1385`)
  - So we can add this parameter more safely.

Reply:
Yes, but it is you who added the paramteter zc in line 1436, not the patch you backported. 
The backported patch only `adds builtInFS.getPath()` and changes the line breaks.
So why did you add zc in line 1436? I think you should not add it there, and then you don't need the other additional changes either.

Reply:
@GoeLin 

- See the original commit at https://github.com/openjdk/jdk/commit/717792c3b728584413572e7aede83290779be2a2
- in the `jdk` code, the `zc)` parameter was there already in the baseline, yes the back port diff did not add `zc`, but it was there

```diff
-        Files.readAttributes(file.toPath(), BasicFileAttributes.class),
-        zc);
+        Files.readAttributes(builtInFS.getPath(file.getPath()),
+        BasicFileAttributes.class), zc);
```

- Well the `zc)` is not in the baseline of `jdk11-dev`, should we need to add this parameter, to make it **`keep inline with the latest code logic`** ??

Reply:
@GoeLin ,

Because in the original commit https://github.com/openjdk/jdk/commit/717792c3b728584413572e7aede83290779be2a2 , the `zc` parameter exits in the base line already. so in `jdk11u-dev` we were trying to keep in line with the latest code base.

Reply:
Ok, if the reason is only to get 11 closer to 17, I think it is better to remove zc at the call site and in the callee from this change here.  The code belongs to [JDK-8243254](https://bugs.openjdk.org/browse/JDK-8243254), but that change is an enhancement and quite large, so it's not a good candidate for backporting.
Here we just want to fix the bug addressed.